File: webclient/src/config/SpotAnimType.ts
import { ConfigType } from '#/config/ConfigType.js';
import SeqType from '#/config/SeqType.js';

import LruCache from '#/datastruct/LruCache.js';

import Model from '#/graphics/Model.js';

import Jagfile from '#/io/Jagfile.js';
import Packet from '#/io/Packet.js';

export default class SpotAnimType extends ConfigType {
    static totalCount: number = 0;
    static instances: SpotAnimType[] = [];
    static modelCache: LruCache | null = new LruCache(30);

    static unpack(config: Jagfile): void {
        const dat: Packet = new Packet(config.read('spotanim.dat'));
        this.totalCount = dat.g2();
        for (let i: number = 0; i < this.totalCount; i++) {
            this.instances[i] = new SpotAnimType(i).unpackType(dat);
        }
    }

    // ----

    model: number = 0;
    anim: number = -1;
    seq: SeqType | null = null;
    disposeAlpha: boolean = false;
    recol_s: Uint16Array = new Uint16Array(6);
    recol_d: Uint16Array = new Uint16Array(6);
    resizeh: number = 128;
    resizev: number = 128;
    spotAngle: number = 0;
    ambient: number = 0;
    contrast: number = 0;

    unpack(code: number, dat: Packet): void {
        if (code === 1) {
            this.model = dat.g2();
        } else if (code === 2) {
            this.anim = dat.g2();

            if (SeqType.instances) {
                this.seq = SeqType.instances[this.anim];
            }
        } else if (code === 3) {
            this.disposeAlpha = true;
        } else if (code === 4) {
            this.resizeh = dat.g2();
        } else if (code === 5) {
            this.resizev = dat.g2();
        } else if (code === 6) {
            this.spotAngle = dat.g2();
        } else if (code === 7) {
            this.ambient = dat.g1();
        } else if (code === 8) {
            this.contrast = dat.g1();
        } else if (code >= 40 && code < 50) {
            this.recol_s[code - 40] = dat.g2();
        } else if (code >= 50 && code < 60) {
            this.recol_d[code - 50] = dat.g2();
        } else {
            console.log('Error unrecognised spotanim config code: ', code);
        }
    }

    getModel(): Model {
        let model: Model | null = SpotAnimType.modelCache?.get(BigInt(this.id)) as Model | null;
        if (model) {
            return model;
        }
        model = Model.model(this.model);
        for (let i: number = 0; i < 6; i++) {
            if (this.recol_s[0] !== 0) {
                model.recolor(this.recol_s[i], this.recol_d[i]);
            }
        }
        SpotAnimType.modelCache?.put(BigInt(this.id), model);
        return model;
    }
}

File: javaclient/src/main/java/jagex2/config/SpotAnimType.java
package jagex2.config;

import deob.ObfuscatedName;
import jagex2.dash3d.Model;
import jagex2.datastruct.LruCache;
import jagex2.io.Jagfile;
import jagex2.io.Packet;

@ObfuscatedName("oc")
public class SpotAnimType {

	@ObfuscatedName("oc.a")
	public static int count;

	@ObfuscatedName("oc.b")
	public static SpotAnimType[] types;

	@ObfuscatedName("oc.c")
	public int id;

	@ObfuscatedName("oc.d")
	public int model;

	@ObfuscatedName("oc.e")
	public int anim = -1;

	@ObfuscatedName("oc.f")
	public SeqType seq;

	@ObfuscatedName("oc.g")
	public boolean animHasAlpha = false;

	@ObfuscatedName("oc.h")
	public int[] recol_s = new int[6];

	@ObfuscatedName("oc.i")
	public int[] recol_d = new int[6];

	@ObfuscatedName("oc.j")
	public int resizeh = 128;

	@ObfuscatedName("oc.k")
	public int resizev = 128;

	@ObfuscatedName("oc.l")
	public int angle;

	@ObfuscatedName("oc.m")
	public int ambient;

	@ObfuscatedName("oc.n")
	public int contrast;

	@ObfuscatedName("oc.o")
	public static LruCache modelCache = new LruCache(30);

	@ObfuscatedName("oc.a(ILyb;)V")
	public static void unpack(Jagfile arg1) {
		Packet var2 = new Packet(arg1.read("spotanim.dat", null));
		count = var2.g2();
		if (types == null) {
			types = new SpotAnimType[count];
		}
		for (int var4 = 0; var4 < count; var4++) {
			if (types[var4] == null) {
				types[var4] = new SpotAnimType();
			}
			types[var4].id = var4;
			types[var4].decode(var2);
		}
	}

	@ObfuscatedName("oc.a(ZLmb;)V")
	public void decode(Packet arg1) {
		while (true) {
			int var4 = arg1.g1();
			if (var4 == 0) {
				return;
			}
			if (var4 == 1) {
				this.model = arg1.g2();
			} else if (var4 == 2) {
				this.anim = arg1.g2();
				if (SeqType.types != null) {
					this.seq = SeqType.types[this.anim];
				}
			} else if (var4 == 3) {
				this.animHasAlpha = true;
			} else if (var4 == 4) {
				this.resizeh = arg1.g2();
			} else if (var4 == 5) {
				this.resizev = arg1.g2();
			} else if (var4 == 6) {
				this.angle = arg1.g2();
			} else if (var4 == 7) {
				this.ambient = arg1.g1();
			} else if (var4 == 8) {
				this.contrast = arg1.g1();
			} else if (var4 >= 40 && var4 < 50) {
				this.recol_s[var4 - 40] = arg1.g2();
			} else if (var4 >= 50 && var4 < 60) {
				this.recol_d[var4 - 50] = arg1.g2();
			} else {
				System.out.println("Error unrecognised spotanim config code: " + var4);
			}
		}
	}

	@ObfuscatedName("oc.a()Lfb;")
	public Model getModel() {
		Model var1 = (Model) modelCache.get((long) this.id);
		if (var1 != null) {
			return var1;
		}
		Model var2 = Model.tryGet(this.model);
		if (var2 == null) {
			return null;
		}
		for (int var3 = 0; var3 < 6; var3++) {
			if (this.recol_s[0] != 0) {
				var2.recolour(this.recol_s[var3], this.recol_d[var3]);
			}
		}
		modelCache.put(var2, (long) this.id);
		return var2;
	}
}
